x/airdrop/keeper/msg_server.go

```diff
@@ -0,0 +1,148 @@
+package keeper
+
+import (
+	"context"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/Stride-Labs/stride/v22/x/airdrop/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+// User transaction to claim all the pending airdrop rewards up to the current day
+func (ms msgServer) ClaimDaily(goCtx context.Context, msg *types.MsgClaimDaily) (*types.MsgClaimDailyResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	err := ms.Keeper.ClaimDaily(ctx, msg.AirdropId, msg.Claimer)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.MsgClaimDailyResponse{}, nil
+}
+
+// User transaction to claim half of their total amount now, and forfeit the other half to be clawed back
+func (ms msgServer) ClaimEarly(goCtx context.Context, msg *types.MsgClaimEarly) (*types.MsgClaimEarlyResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	err := ms.Keeper.ClaimEarly(ctx, msg.AirdropId, msg.Claimer)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.MsgClaimEarlyResponse{}, nil
+}
+
+// Admin transaction to create a new airdrop
```

Comment:
Will we ever want to create an airdrop through the admin function? Or is this mainly "just-in-case"?

Reply:
yeah this is actually how we're planning to do it! (not via the upgrade handler)

Reply:
Oh gotcha! Is that feasible? If there are 10k+ addresses, do you know if that's 100 txs, 10 txs, etc?

Reply:
yeah we should test it - there's roughly 65k addresses in the airdrop. 

The way we were thinking about it was that it was a tradeoff between:
* being forced to run an upgrade to add allocations (e.g. we'd need to do another upgrade for stDYM) 
* vs doing it manually and having less confidence that all the txs went through

But I think we can mitigate the second point by just auditing the chain data after we add everything.

We should try and see how many we can get in a single message though. It's on my docket to test, but if you can get to to it before me that'd be great!

Reply:
tested offline and confirmed it will probably be on the order of ~30 txs (assuming not submitted by a ledger admin)

marking resolved!